content/blog/2024/2024-06-06-junit5-functional-interfaces.md

```diff
@@ -0,0 +1,620 @@
+
+## JUnit 5 Functional Interfaces
+
+JUnit functional interfaces belong to `org.junit.jupiter.api.function` package. **It defines three functional interfaces: `Executable`, `ThrowingConsumer<T>` and `ThrowingSupplier<T>`.**  We use them typically with `Assertions` utility class.
+
+Knowing how these interfaces work can greatly improve your testing methods. They make it easier to write and comprehend tests, reducing the amount of repetitive code needed for handling exceptions. By using these interfaces, you can describe complex test scenarios more clearly and concisely.
+
+Let's learn how to use these function interfaces.
+
+## Using `Executable`
+
+**The `Executable` is a functional interface that enables the implementation of any generic block of code that may potentially throw a `Throwable`.**
+
+JUnit 5 defines the `Executable` interface as follows::
+
+```java
+@FunctionalInterface
+public interface Executable {
+    void execute() throws Throwable;
+}
+```
+
+**The `Executable` interface defines a single method called execute, which does not have any parameters and does not return anything.** It can throw different types of exceptions, making it flexible for handling exceptional situations.
+
+It is particularly useful for writing tests to validate if specific code paths throw specific exceptions.
+
+One common scenario where we use the `Executable` with the `Assertions.assertThrows()` method. This method takes a `Executable` as an argument, executes it, and checks if it throws the expected exception.
+
+[`Assertions`](https://junit.org/junit5/docs/5.9.0/api/org.junit.jupiter.api/org/junit/jupiter/api/Assertions.html) class has many assertion methods accepting executable:
+
+```java
+static void assertAll(Executable... executables)
+// other variants of assertAll accepting executables
+
+static void assertDoesNotThrow(Executable executable)
+// other variants of assertDoesNotThrow accepting executables
+
+static <T extends Throwable> T assertThrows(Class<T> expectedType,
+                                            Executable executable)
+// other variants of assertThrows accepting executables
+
+static <T extends Throwable> T assertThrowsExactly(Class<T> expectedType,
+                                                   Executable executable)
+// other variants of assertThrowsExactly accepting executables
+
+static void assertTimeout(Duration timeout, Executable executable)
+// other variants of assertTimeout accepting executables
+
+static void assertTimeoutPreemptively(Duration timeout, Executable executable)
+// other variants of assertTimeoutPreemptively accepting executables
+```
+
+Let's learn about these methods one by one.
+
+### Using Executables in `assertAll()`
+
+**The `Assertions.assertsAll()` method asserts that all supplied executables do not throw exceptions.**
+
+Let's now see how to use `assertAll()` and executables:
+
+```java
```

Comment:
We introduce the below code snippet saying that we'll use `assertAll()`. However, the code snippet does not contain this method, which is a little confusing at first glance. Let's try to improve this.

Reply:
AssertAll is used in next para. Updated it,